### PR TITLE
feat: dual publish mode build target and integration support

### PR DESCRIPTION
**Pull request by barak007**
### cli

- better esm output in cli
- add side effects css imports to generated module

### core 

- `tryCollectImportsDeep` also calc css depth
- `tryCollectImportsDeep` `onImport` hook

### webpack 

- add `cssDepth` 
- bug fix in `calcDepth` can pick up files like `1st.css1`
- added depthStrategy

### test kit

- add options to build test packages.
- add ability to provide array of webpack configs.
- open browser in internal path.

### CODE REVIEW

File: packages/cli/src/build-single-file.ts [hunk 1]
```diff
@@ -135,22 +136,41 @@ export function buildSingleFile({
...
     // st.css.js
     moduleFormats.forEach((format) => {
+        const ext = format === 'esm' ? '.mjs' : '.js';
         outputLogs.push(`${format} module`);
+
+        const moduleCssImports = injectCSSRequest
+            ? [moduleRequestSourceCode(format, './' + cssAssetFilename)]
+            : [];
+        let cssDepth = -1;
+        tryCollectImportsDeep(stylable, res.meta, new Set(), ({ depth }) => {
```
Comment (barak007) [inline packages/cli/src/build-single-file.ts:146]: css depth only

File: packages/cli/src/build-single-file.ts [hunk 2]
```diff
@@ -135,22 +136,41 @@ export function buildSingleFile({
...
+            cssDepth = Math.max(cssDepth, depth);
+        });
+        for (const imported of res.meta.getImportStatements()) {
+            if (imported.request.endsWith('.st.css')) {
```
Comment (barak007) [inline packages/cli/src/build-single-file.ts:150]: needs to be resolved

File: packages/core/src/helpers/import.ts [hunk 3]
```diff
@@ -580,18 +580,36 @@ function isImportAs(space: ParsedValue, as: ParsedValue) {
...
-    imports = new Set<string>()
+    imports = new Set<string>(),
+    onImport: undefined | ((e: ImportEvent) => void) = undefined,
+    depth = 0
 ) {
     for (const { context, request } of meta.getImportStatements()) {
         try {
             const resolved = stylable.resolver.resolvePath(context, request);
+            if (onImport) {
+                onImport({ context, request, resolved, depth });
```
Comment (barak007) [inline packages/core/src/helpers/import.ts:601]: `?.`

File: packages/webpack-plugin/src/plugin-utils.ts [hunk 4]
```diff
@@ -479,10 +484,14 @@ export function createCalcDepthContext(moduleGraph: ModuleGraph): CalcDepthConte
             ),
         getModulePathNoExt: (module) => {
             if (isStylableModule(module)) {
-                return module.resource.replace(/\.st\.css/g, '');
+                return module.resource.replace(/\.st\.css$/, '');
```
Comment (barak007) [inline packages/webpack-plugin/src/plugin-utils.ts:487]: bugfix

File: packages/webpack-plugin/src/plugin-utils.ts [hunk 5]
```diff
@@ -479,10 +484,14 @@ export function createCalcDepthContext(moduleGraph: ModuleGraph): CalcDepthConte
...
             }
             const { dir, name } = parse((module as NormalModule)?.resource || '');
-            return join(dir, name);
+            let finalName = name;
```
Comment (barak007) [inline packages/webpack-plugin/src/plugin-utils.ts:490]: not sure why this code is here. needs double check

File: packages/webpack-plugin/test/e2e/dual-mode-esm.spec.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    it('output same styles depth with both integrations', async () => {
+        const vanillaBrowser = await projectRunner.openInBrowser({ internalPath: 'vanilla' });
+        const vanillaStyles = await vanillaBrowser.page.evaluate(getStyleElementsMetadata);
+        const stylableBrowser = await projectRunner.openInBrowser({ internalPath: 'stylable' });
+        const stylableStyles = await stylableBrowser.page.evaluate(getStyleElementsMetadata);
+
+        expect(vanillaStyles).to.eql(stylableStyles);
+
+        expect(normalizeNamespace(vanillaStyles)).to.eql([
```
Comment (idoros) [inline packages/webpack-plugin/test/e2e/dual-mode-esm.spec.ts:35]: add computed style assertions

File: packages/cli/src/build.ts [hunk 7]
```diff
@@ -236,6 +275,36 @@ export async function build(
                     dtsSourceMap,
                     minify,
                     generated,
+                    resolveRuntimeRequest: (targetFilePath, moduleFormat) => {
+                        if (inlineRuntime) {
+                            if (moduleFormat === 'cjs' && runtimeCjsOutPath) {
```
Comment (barak007) [inline packages/cli/src/build.ts:280]: refactor

File: packages/cli/src/build.ts [hunk 8]
```diff
@@ -312,6 +381,96 @@ export async function build(
...
+            fixRelativeUrls(ast, targetFilePath, join(fullOutDir, bundle));
+            return ast.toString();
+        })
+        .join('\n');
+
+    const optimizer = new StylableOptimizer();
+    return [
+        {
+            filePath: join(fullOutDir, bundle),
+            content: minify ? optimizer.minifyCSS(cssBundleCode) : cssBundleCode,
```
Comment (barak007) [inline packages/cli/src/build.ts:426]: inline the new StylableOptimizer

File: packages/cli/src/build.ts [hunk 9]
```diff
@@ -312,6 +381,96 @@ export async function build(
...
+            : require.resolve(runtimeCjsRequest, {
+                  paths: [projectRoot],
+              });
+        const runtimeEsmPath = fs.isAbsolute(runtimeEsmRequest)
+            ? runtimeEsmRequest
+            : require.resolve(runtimeEsmRequest, {
+                  paths: [projectRoot],
+              });
+
+        // TODO: inline the inject styles. done in the #2615
```
Comment (barak007) [inline packages/cli/src/build.ts:457]: this was added to the pure.ts. remove todo

File: packages/core/src/stylable-transformer.ts [hunk 10]
```diff
@@ -543,6 +545,43 @@ export class StylableTransformer {
             });
         }
     }
+    private calcCssDepth(meta: StylableMeta) {
```
Comment (barak007) [inline packages/core/src/stylable-transformer.ts:548]: move to StImport

File: packages/core/src/stylable-transformer.ts [hunk 11]
```diff
@@ -543,6 +545,43 @@ export class StylableTransformer {
...
+            meta,
+            new Set(),
+            ({ depth }) => {
+                cssDepth = Math.max(cssDepth, depth);
+            },
+            1
+        );
+        return { cssDepth, deepDependencies };
+    }
+    private isDuplicateStScopeDiagnostic(context: ScopeContext) {
```
Comment (barak007) [inline packages/core/src/stylable-transformer.ts:564]: remove

File: packages/core/src/stylable-transformer.ts [hunk 12]
```diff
@@ -543,6 +545,43 @@ export class StylableTransformer {
             });
         }
     }
+    private calcCssDepth(meta: StylableMeta) {
+        let cssDepth = 0;
+        const deepDependencies = tryCollectImportsDeep(
```
Comment (barak007) [inline packages/core/src/stylable-transformer.ts:550]: remove deepDependencies

File: packages/webpack-plugin/src/plugin-utils.ts [hunk 13]
```diff
@@ -208,6 +211,15 @@ export function injectLoader(compiler: Compiler) {
         loader: LOADER_NAME,
         sideEffects: true,
     });
+    options.resolve ||= {};
+    options.resolve.plugins ||= [];
+
+    options.resolve.plugins.push(new ReExt(/./, /\.st\.css\.js$/, '.st.css'));
```
Comment (barak007) [inline packages/webpack-plugin/src/plugin-utils.ts:217]: improve regexp to support mjs cjs

Comment (barak007) [reply packages/webpack-plugin/src/plugin-utils.ts:217]: remove comment and loader code

File: packages/webpack-plugin/src/plugin-utils.ts (hunk 5 above)
Comment (barak007) [reply packages/webpack-plugin/src/plugin-utils.ts:517]: re direct view can we remove this?

File: packages/webpack-plugin/src/plugin-utils.ts [hunk 14]
```diff
@@ -584,3 +600,65 @@ export function isDependencyOf(entryPoint: EntryPoint, entrypoints: Iterable<Ent
     }
     return false;
 }
+
+class ReExt {
```
Comment (barak007) [inline packages/webpack-plugin/src/plugin-utils.ts:604]: move to new file

File: packages/webpack-plugin/src/redirect-to-source-loader.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import type { LoaderContext } from 'webpack';
```
Comment (barak007) [inline packages/webpack-plugin/src/redirect-to-source-loader.ts:1]: delete

File: packages/build-tools/src/calc-depth.ts [hunk 16]
```diff
@@ -42,16 +42,18 @@ export function calcDepth<T>(
...
         const parentViewsList: T[] = [];
         const viewPath = context.getModulePathNoExt(module);
         for (const importer of context.getImporters(module)) {
             if (
                 !context.isStylableModule(importer) &&
                 context.getModulePathNoExt(importer) === viewPath
             ) {
-                parentViewsList.push(importer);
+                // try to find the view from a redirected js request
+                const redirectedView = inner ? false : getCSSViewModule(importer, context, true);
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/build-tools/src/calc-depth.ts:55]: yes, removed

File: packages/core/src/stylable-transformer.ts (hunk 12 above)
Comment (barak007) [reply packages/core/src/stylable-transformer.ts:550]: we are going to keep them. used to report build dependencies

File: packages/cli/src/build-single-file.ts [hunk 17]
```diff
@@ -290,6 +299,23 @@ export function buildSingleFile({
...
 }
 
+function inlineAssetsForJsModule(res: StylableResults, stylable: Stylable, fs: IFileSystem) {
+    const ast = res.meta.targetAst!.clone();
+    processUrlDependencies({
+        meta: { targetAst: ast, source: res.meta.source },
+        rootContext: stylable.projectRoot,
+        getReplacement: ({ absoluteRequest, url }) => {
+            if (isAsset(url)) {
+                const content = fs.readFileSync(absoluteRequest);
```
Comment (barak007) [inline packages/cli/src/build-single-file.ts:309]: handle ~

File: packages/experimental-loader/package.json [hunk 18]
```diff
@@ -9,7 +9,7 @@
     "@stylable/core": "^5.4.0",
     "@stylable/optimizer": "^5.4.0",
     "@stylable/runtime": "^5.4.0",
-    "css-loader": "^6.7.2",
+    "css-loader": "^6.7.1",
```
Comment (barak007) [inline packages/experimental-loader/package.json:12]:
```suggestion
    "css-loader": "^6.7.2",
```

**Review by idoros — APPROVED** [review]

**Pull request merged**